// @tbd54566975/dwn-sdk-js/dist/cjs/src/interfaces/records/handlers/records-delete.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructIndexes = exports.RecordsDeleteHandler = void 0;
var auth_js_1 = require("../../../core/auth.js");
var records_interface_js_1 = require("../records-interface.js");
var message_reply_js_1 = require("../../../core/message-reply.js");
var records_delete_js_1 = require("../messages/records-delete.js");
var records_write_js_1 = require("../messages/records-write.js");
var message_js_1 = require("../../../core/message.js");
var RecordsDeleteHandler = /** @class */ (function () {
    function RecordsDeleteHandler(didResolver, messageStore, dataStore, eventLog) {
        this.didResolver = didResolver;
        this.messageStore = messageStore;
        this.dataStore = dataStore;
        this.eventLog = eventLog;
    }
    RecordsDeleteHandler.prototype.handle = function (_a) {
        var tenant = _a.tenant, message = _a.message;
        return __awaiter(this, void 0, void 0, function () {
            var recordsDelete, e_1, e_2, query, existingMessages, newestExistingMessage, incomingMessageIsNewest, newestMessage, _b, indexes, messageCid, messageReply;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, records_delete_js_1.RecordsDelete.parse(message)];
                    case 1:
                        recordsDelete = _c.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _c.sent();
                        return [2 /*return*/, message_reply_js_1.MessageReply.fromError(e_1, 400)];
                    case 3:
                        _c.trys.push([3, 6, , 7]);
                        return [4 /*yield*/, (0, auth_js_1.authenticate)(message.authorization, this.didResolver)];
                    case 4:
                        _c.sent();
                        return [4 /*yield*/, recordsDelete.authorize(tenant)];
                    case 5:
                        _c.sent();
                        return [3 /*break*/, 7];
                    case 6:
                        e_2 = _c.sent();
                        return [2 /*return*/, message_reply_js_1.MessageReply.fromError(e_2, 401)];
                    case 7:
                        query = {
                            interface: message_js_1.DwnInterfaceName.Records,
                            recordId: message.descriptor.recordId
                        };
                        return [4 /*yield*/, this.messageStore.query(tenant, query)];
                    case 8:
                        existingMessages = _c.sent();
                        return [4 /*yield*/, records_write_js_1.RecordsWrite.getNewestMessage(existingMessages)];
                    case 9:
                        newestExistingMessage = _c.sent();
                        incomingMessageIsNewest = false;
                        _b = newestExistingMessage === undefined;
                        if (_b) return [3 /*break*/, 11];
                        return [4 /*yield*/, records_write_js_1.RecordsWrite.isNewer(message, newestExistingMessage)];
                    case 10:
                        _b = (_c.sent());
                        _c.label = 11;
                    case 11:
                        // if incoming message is newest
                        if (_b) {
                            incomingMessageIsNewest = true;
                            newestMessage = message;
                        }
                        else { // existing message is the same age or newer than the incoming message
                            newestMessage = newestExistingMessage;
                        }
                        if (!incomingMessageIsNewest) {
                            return [2 /*return*/, new message_reply_js_1.MessageReply({
                                    status: { code: 409, detail: 'Conflict' }
                                })];
                        }
                        // return Not Found if record does not exist or is already deleted
                        if (newestExistingMessage === undefined || newestExistingMessage.descriptor.method === message_js_1.DwnMethodName.Delete) {
                            return [2 /*return*/, new message_reply_js_1.MessageReply({
                                    status: { code: 404, detail: 'Not Found' }
                                })];
                        }
                        return [4 /*yield*/, constructIndexes(tenant, recordsDelete)];
                    case 12:
                        indexes = _c.sent();
                        return [4 /*yield*/, this.messageStore.put(tenant, message, indexes)];
                    case 13:
                        _c.sent();
                        return [4 /*yield*/, message_js_1.Message.getCid(message)];
                    case 14:
                        messageCid = _c.sent();
                        return [4 /*yield*/, this.eventLog.append(tenant, messageCid)];
                    case 15:
                        _c.sent();
                        // delete all existing messages that are not newest, except for the initial write
                        return [4 /*yield*/, (0, records_interface_js_1.deleteAllOlderMessagesButKeepInitialWrite)(tenant, existingMessages, newestMessage, this.messageStore, this.dataStore, this.eventLog)];
                    case 16:
                        // delete all existing messages that are not newest, except for the initial write
                        _c.sent();
                        messageReply = new message_reply_js_1.MessageReply({
                            status: { code: 202, detail: 'Accepted' }
                        });
                        return [2 /*return*/, messageReply];
                }
            });
        });
    };
    ;
    return RecordsDeleteHandler;
}());
exports.RecordsDeleteHandler = RecordsDeleteHandler;
function constructIndexes(tenant, recordsDelete) {
    return __awaiter(this, void 0, void 0, function () {
        var message, descriptor, indexes;
        return __generator(this, function (_a) {
            message = recordsDelete.message;
            descriptor = __assign({}, message.descriptor);
            indexes = __assign({ 
                // isLatestBaseState : "true", // intentionally showing that this index is omitted
                author: recordsDelete.author }, descriptor);
            return [2 /*return*/, indexes];
        });
    });
}
exports.constructIndexes = constructIndexes;
