// @tbd54566975/dwn-sdk-js/dist/esm/src/interfaces/protocols/handlers/protocols-configure.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { canonicalAuth } from '../../../core/auth.js';
import { MessageReply } from '../../../core/message-reply.js';
import { ProtocolsConfigure } from '../messages/protocols-configure.js';
import { StorageController } from '../../../store/storage-controller.js';
import { DwnInterfaceName, DwnMethodName, Message } from '../../../core/message.js';
export class ProtocolsConfigureHandler {
    constructor(didResolver, messageStore, dataStore, eventLog) {
        this.didResolver = didResolver;
        this.messageStore = messageStore;
        this.dataStore = dataStore;
        this.eventLog = eventLog;
    }
    handle({ tenant, message, dataStream: _dataStream }) {
        return __awaiter(this, void 0, void 0, function* () {
            let protocolsConfigure;
            try {
                protocolsConfigure = yield ProtocolsConfigure.parse(message);
            }
            catch (e) {
                return MessageReply.fromError(e, 400);
            }
            // authentication & authorization
            try {
                yield canonicalAuth(tenant, protocolsConfigure, this.didResolver);
            }
            catch (e) {
                return MessageReply.fromError(e, 401);
            }
            // attempt to get existing protocol
            const query = {
                interface: DwnInterfaceName.Protocols,
                method: DwnMethodName.Configure,
                protocol: message.descriptor.definition.protocol
            };
            const existingMessages = yield this.messageStore.query(tenant, query);
            // find lexicographically the largest message, and if the incoming message is the largest
            let newestMessage = yield Message.getMessageWithLargestCid(existingMessages);
            let incomingMessageIsNewest = false;
            if (newestMessage === undefined || (yield Message.isCidLarger(message, newestMessage))) {
                incomingMessageIsNewest = true;
                newestMessage = message;
            }
            // write the incoming message to DB if incoming message is newest
            let messageReply;
            if (incomingMessageIsNewest) {
                const indexes = ProtocolsConfigureHandler.constructProtocolsConfigureIndexes(protocolsConfigure);
                const messageCid = yield Message.getCid(message);
                yield this.messageStore.put(tenant, message, indexes);
                yield this.eventLog.append(tenant, messageCid);
                messageReply = new MessageReply({
                    status: { code: 202, detail: 'Accepted' }
                });
            }
            else {
                messageReply = new MessageReply({
                    status: { code: 409, detail: 'Conflict' }
                });
            }
            // delete all existing records that are smaller
            const deletedMessageCids = [];
            for (const message of existingMessages) {
                if (yield Message.isCidLarger(newestMessage, message)) {
                    const messageCid = yield Message.getCid(message);
                    deletedMessageCids.push(messageCid);
                    yield StorageController.delete(this.messageStore, this.dataStore, tenant, message);
                }
            }
            yield this.eventLog.deleteEventsByCid(tenant, deletedMessageCids);
            return messageReply;
        });
    }
    ;
    static constructProtocolsConfigureIndexes(protocolsConfigure) {
        // strip out `dataSize` and `definition` as they are not indexable
        // retain protocol url from `definition`
        const _a = protocolsConfigure.message.descriptor, { dataSize, definition } = _a, propertiesToIndex = __rest(_a, ["dataSize", "definition"]);
        const { author } = protocolsConfigure;
        const indexes = Object.assign(Object.assign({}, propertiesToIndex), { protocol: definition.protocol, author: author });
        return indexes;
    }
}
