// @tbd54566975/dwn-sdk-js/dist/cjs/src/interfaces/protocols/handlers/protocols-configure.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtocolsConfigureHandler = void 0;
var auth_js_1 = require("../../../core/auth.js");
var message_reply_js_1 = require("../../../core/message-reply.js");
var protocols_configure_js_1 = require("../messages/protocols-configure.js");
var storage_controller_js_1 = require("../../../store/storage-controller.js");
var message_js_1 = require("../../../core/message.js");
var ProtocolsConfigureHandler = /** @class */ (function () {
    function ProtocolsConfigureHandler(didResolver, messageStore, dataStore, eventLog) {
        this.didResolver = didResolver;
        this.messageStore = messageStore;
        this.dataStore = dataStore;
        this.eventLog = eventLog;
    }
    ProtocolsConfigureHandler.prototype.handle = function (_a) {
        var tenant = _a.tenant, message = _a.message, _dataStream = _a.dataStream;
        return __awaiter(this, void 0, void 0, function () {
            var protocolsConfigure, e_1, e_2, query, existingMessages, newestMessage, incomingMessageIsNewest, _b, messageReply, indexes, messageCid, deletedMessageCids, existingMessages_1, existingMessages_1_1, message_1, messageCid, e_3_1;
            var e_3, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, protocols_configure_js_1.ProtocolsConfigure.parse(message)];
                    case 1:
                        protocolsConfigure = _d.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _d.sent();
                        return [2 /*return*/, message_reply_js_1.MessageReply.fromError(e_1, 400)];
                    case 3:
                        _d.trys.push([3, 5, , 6]);
                        return [4 /*yield*/, (0, auth_js_1.canonicalAuth)(tenant, protocolsConfigure, this.didResolver)];
                    case 4:
                        _d.sent();
                        return [3 /*break*/, 6];
                    case 5:
                        e_2 = _d.sent();
                        return [2 /*return*/, message_reply_js_1.MessageReply.fromError(e_2, 401)];
                    case 6:
                        query = {
                            interface: message_js_1.DwnInterfaceName.Protocols,
                            method: message_js_1.DwnMethodName.Configure,
                            protocol: message.descriptor.definition.protocol
                        };
                        return [4 /*yield*/, this.messageStore.query(tenant, query)];
                    case 7:
                        existingMessages = _d.sent();
                        return [4 /*yield*/, message_js_1.Message.getMessageWithLargestCid(existingMessages)];
                    case 8:
                        newestMessage = _d.sent();
                        incomingMessageIsNewest = false;
                        _b = newestMessage === undefined;
                        if (_b) return [3 /*break*/, 10];
                        return [4 /*yield*/, message_js_1.Message.isCidLarger(message, newestMessage)];
                    case 9:
                        _b = (_d.sent());
                        _d.label = 10;
                    case 10:
                        if (_b) {
                            incomingMessageIsNewest = true;
                            newestMessage = message;
                        }
                        if (!incomingMessageIsNewest) return [3 /*break*/, 14];
                        indexes = ProtocolsConfigureHandler.constructProtocolsConfigureIndexes(protocolsConfigure);
                        return [4 /*yield*/, message_js_1.Message.getCid(message)];
                    case 11:
                        messageCid = _d.sent();
                        return [4 /*yield*/, this.messageStore.put(tenant, message, indexes)];
                    case 12:
                        _d.sent();
                        return [4 /*yield*/, this.eventLog.append(tenant, messageCid)];
                    case 13:
                        _d.sent();
                        messageReply = new message_reply_js_1.MessageReply({
                            status: { code: 202, detail: 'Accepted' }
                        });
                        return [3 /*break*/, 15];
                    case 14:
                        messageReply = new message_reply_js_1.MessageReply({
                            status: { code: 409, detail: 'Conflict' }
                        });
                        _d.label = 15;
                    case 15:
                        deletedMessageCids = [];
                        _d.label = 16;
                    case 16:
                        _d.trys.push([16, 23, 24, 25]);
                        existingMessages_1 = __values(existingMessages), existingMessages_1_1 = existingMessages_1.next();
                        _d.label = 17;
                    case 17:
                        if (!!existingMessages_1_1.done) return [3 /*break*/, 22];
                        message_1 = existingMessages_1_1.value;
                        return [4 /*yield*/, message_js_1.Message.isCidLarger(newestMessage, message_1)];
                    case 18:
                        if (!_d.sent()) return [3 /*break*/, 21];
                        return [4 /*yield*/, message_js_1.Message.getCid(message_1)];
                    case 19:
                        messageCid = _d.sent();
                        deletedMessageCids.push(messageCid);
                        return [4 /*yield*/, storage_controller_js_1.StorageController.delete(this.messageStore, this.dataStore, tenant, message_1)];
                    case 20:
                        _d.sent();
                        _d.label = 21;
                    case 21:
                        existingMessages_1_1 = existingMessages_1.next();
                        return [3 /*break*/, 17];
                    case 22: return [3 /*break*/, 25];
                    case 23:
                        e_3_1 = _d.sent();
                        e_3 = { error: e_3_1 };
                        return [3 /*break*/, 25];
                    case 24:
                        try {
                            if (existingMessages_1_1 && !existingMessages_1_1.done && (_c = existingMessages_1.return)) _c.call(existingMessages_1);
                        }
                        finally { if (e_3) throw e_3.error; }
                        return [7 /*endfinally*/];
                    case 25: return [4 /*yield*/, this.eventLog.deleteEventsByCid(tenant, deletedMessageCids)];
                    case 26:
                        _d.sent();
                        return [2 /*return*/, messageReply];
                }
            });
        });
    };
    ;
    ProtocolsConfigureHandler.constructProtocolsConfigureIndexes = function (protocolsConfigure) {
        // strip out `dataSize` and `definition` as they are not indexable
        // retain protocol url from `definition`
        var _a = protocolsConfigure.message.descriptor, dataSize = _a.dataSize, definition = _a.definition, propertiesToIndex = __rest(_a, ["dataSize", "definition"]);
        var author = protocolsConfigure.author;
        var indexes = __assign(__assign({}, propertiesToIndex), { protocol: definition.protocol, author: author });
        return indexes;
    };
    return ProtocolsConfigureHandler;
}());
exports.ProtocolsConfigureHandler = ProtocolsConfigureHandler;
