// @decentralized-identity/ion-sdk/dist/lib/IonKey.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Ed25519 = require("@noble/ed25519");
const Secp256k1 = require("@noble/secp256k1");
const InputValidator_1 = require("./InputValidator");
const base64_1 = require("multiformats/bases/base64");
/**
 * Class containing operations related to keys used in ION.
 */
class IonKey {
    /**
     * Generates SECP256K1 key pair to be used in an operation.
     * Mainly used for testing.
     * @returns [publicKey, privateKey]
     */
    static generateEs256kDidDocumentKeyPair(input) {
        return __awaiter(this, void 0, void 0, function* () {
            const id = input.id;
            const purposes = input.purposes;
            InputValidator_1.default.validateId(id);
            InputValidator_1.default.validatePublicKeyPurposes(purposes);
            const [publicKey, privateKey] = yield IonKey.generateEs256kKeyPair();
            const publicKeyModel = {
                id,
                type: 'EcdsaSecp256k1VerificationKey2019',
                publicKeyJwk: publicKey
            };
            // Only add the `purposes` property If given `purposes` array has at least an entry.
            if (purposes !== undefined && purposes.length > 0) {
                publicKeyModel.purposes = purposes;
            }
            return [publicKeyModel, privateKey];
        });
    }
    /**
     * Generates SECP256K1 key pair for ION operation use.
     * @returns [publicKey, privateKey]
     */
    static generateEs256kOperationKeyPair() {
        return __awaiter(this, void 0, void 0, function* () {
            const keyPair = yield IonKey.generateEs256kKeyPair();
            return keyPair;
        });
    }
    static generateEs256kKeyPair() {
        return __awaiter(this, void 0, void 0, function* () {
            const privateKeyBytes = Secp256k1.utils.randomPrivateKey();
            // the public key is uncompressed which means that it contains both the x and y values.
            // the first byte is a header that indicates whether the key is uncompressed (0x04 if uncompressed).
            // bytes 1 - 32 represent X
            // bytes 33 - 64 represent Y
            const publicKeyBytes = yield Secp256k1.getPublicKey(privateKeyBytes);
            const d = base64_1.base64url.baseEncode(privateKeyBytes);
            // skip the first byte because it's used as a header to indicate whether the key is uncompressed
            const x = base64_1.base64url.baseEncode(publicKeyBytes.subarray(1, 33));
            const y = base64_1.base64url.baseEncode(publicKeyBytes.subarray(33, 65));
            const publicJwk = {
                // alg: 'ES256K',
                kty: 'EC',
                crv: 'secp256k1',
                x,
                y
            };
            const privateJwk = Object.assign(Object.assign({}, publicJwk), { d });
            return [publicJwk, privateJwk];
        });
    }
    /**
     * Generates Ed25519 key pair to be used in an operation.
     * Mainly used for testing.
     * @returns [publicKey, privateKey]
     */
    static generateEd25519DidDocumentKeyPair(input) {
        return __awaiter(this, void 0, void 0, function* () {
            const id = input.id;
            const purposes = input.purposes;
            InputValidator_1.default.validateId(id);
            InputValidator_1.default.validatePublicKeyPurposes(purposes);
            const [publicKey, privateKey] = yield IonKey.generateEd25519KeyPair();
            const publicKeyModel = {
                id,
                type: 'JsonWebKey2020',
                publicKeyJwk: publicKey
            };
            // Only add the `purposes` property If given `purposes` array has at least an entry.
            if (purposes !== undefined && purposes.length > 0) {
                publicKeyModel.purposes = purposes;
            }
            return [publicKeyModel, privateKey];
        });
    }
    /**
     * Generates Ed25519 key pair for ION operation use.
     * @returns [publicKey, privateKey]
     */
    static generateEd25519OperationKeyPair() {
        return __awaiter(this, void 0, void 0, function* () {
            const keyPair = yield IonKey.generateEd25519KeyPair();
            return keyPair;
        });
    }
    static generateEd25519KeyPair() {
        return __awaiter(this, void 0, void 0, function* () {
            const privateKeyBytes = Ed25519.utils.randomPrivateKey();
            const publicKeyBytes = yield Ed25519.getPublicKey(privateKeyBytes);
            const d = base64_1.base64url.baseEncode(privateKeyBytes);
            const x = base64_1.base64url.baseEncode(publicKeyBytes);
            const publicJwk = {
                // alg: 'EdDSA',
                kty: 'OKP',
                crv: 'Ed25519',
                x
            };
            const privateJwk = Object.assign(Object.assign({}, publicJwk), { d });
            return [publicJwk, privateJwk];
        });
    }
    static isJwkEs256k(key) {
        return key.crv === 'secp256k1' && key.kty === 'EC';
    }
    ;
    static isJwkEd25519(key) {
        return key.crv === 'Ed25519' && key.kty === 'OKP';
    }
    ;
}
exports.default = IonKey;
