// @tbd54566975/dwn-sdk-js/dist/esm/src/store/level-wrapper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
import { abortOr } from '../utils/abort.js';
import { sleep } from '../utils/time.js';
export function createLevelDatabase(location, options) {
    return __awaiter(this, void 0, void 0, function* () {
        // Only import `'level'` when it's actually necessary (i.e. only when the default `createLevelDatabase` is used).
        // Overriding `createLevelDatabase` will prevent this from happening.
        const { Level } = yield import('level');
        return new Level(location, Object.assign(Object.assign({}, options), { keyEncoding: 'utf8' }));
    });
}
// `Level` works in Node.js 12+ and Electron 5+ on Linux, Mac OS, Windows and FreeBSD, including any
// future Node.js and Electron release thanks to Node-API, including ARM platforms like Raspberry Pi
// and Android, as well as in Chrome, Firefox, Edge, Safari, iOS Safari and Chrome for Android.
export class LevelWrapper {
    /**
     * @param config.location - must be a directory path (relative or absolute) where `Level`` will
     * store its files, or in browsers, the name of the {@link https://developer.mozilla.org/en-US/docs/Web/API/IDBDatabase `IDBDatabase`}
     * to be opened.
     */
    constructor(config, db) {
        this.config = Object.assign({ createLevelDatabase }, config);
        this.db = db;
    }
    open() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.createLevelDatabase();
            while (this.db.status === 'opening' || this.db.status === 'closing') {
                yield sleep(200);
            }
            if (this.db.status === 'open') {
                return;
            }
            // `db.open()` is automatically called by the database constructor.  We're calling it explicitly
            // in order to explicitly catch an error that would otherwise not surface until another method
            // like `db.get()` is called.  Once `db.open()` has then been called, any read & write
            // operations will again be queued internally until opening has finished.
            return this.db.open();
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.db) {
                return;
            }
            while (this.db.status === 'opening' || this.db.status === 'closing') {
                yield sleep(200);
            }
            if (this.db.status === 'closed') {
                return;
            }
            return this.db.close();
        });
    }
    partition(name) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.createLevelDatabase();
            return new LevelWrapper(this.config, this.db.sublevel(name, {
                keyEncoding: 'utf8',
                valueEncoding: this.config.valueEncoding
            }));
        });
    }
    get(key, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            (_a = options === null || options === void 0 ? void 0 : options.signal) === null || _a === void 0 ? void 0 : _a.throwIfAborted();
            yield abortOr(options === null || options === void 0 ? void 0 : options.signal, this.createLevelDatabase());
            try {
                const value = yield abortOr(options === null || options === void 0 ? void 0 : options.signal, this.db.get(String(key)));
                return value;
            }
            catch (error) {
                const e = error;
                // `Level`` throws an error if the key is not present.  Return `undefined` in this case.
                if (e.code === 'LEVEL_NOT_FOUND') {
                    return undefined;
                }
                else {
                    throw error;
                }
            }
        });
    }
    has(key, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return !!(yield this.get(key, options));
        });
    }
    keys(options) {
        var _a, _b;
        return __asyncGenerator(this, arguments, function* keys_1() {
            var e_1, _c;
            (_a = options === null || options === void 0 ? void 0 : options.signal) === null || _a === void 0 ? void 0 : _a.throwIfAborted();
            yield __await(abortOr(options === null || options === void 0 ? void 0 : options.signal, this.createLevelDatabase()));
            try {
                for (var _d = __asyncValues(this.db.keys()), _e; _e = yield __await(_d.next()), !_e.done;) {
                    const key = _e.value;
                    (_b = options === null || options === void 0 ? void 0 : options.signal) === null || _b === void 0 ? void 0 : _b.throwIfAborted();
                    yield yield __await(key);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_e && !_e.done && (_c = _d.return)) yield __await(_c.call(_d));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    iterator(iteratorOptions, options) {
        var _a, _b;
        return __asyncGenerator(this, arguments, function* iterator_1() {
            var e_2, _c;
            (_a = options === null || options === void 0 ? void 0 : options.signal) === null || _a === void 0 ? void 0 : _a.throwIfAborted();
            yield __await(abortOr(options === null || options === void 0 ? void 0 : options.signal, this.createLevelDatabase()));
            try {
                for (var _d = __asyncValues(this.db.iterator(iteratorOptions)), _e; _e = yield __await(_d.next()), !_e.done;) {
                    const entry = _e.value;
                    (_b = options === null || options === void 0 ? void 0 : options.signal) === null || _b === void 0 ? void 0 : _b.throwIfAborted();
                    yield yield __await(entry);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_e && !_e.done && (_c = _d.return)) yield __await(_c.call(_d));
                }
                finally { if (e_2) throw e_2.error; }
            }
        });
    }
    put(key, value, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            (_a = options === null || options === void 0 ? void 0 : options.signal) === null || _a === void 0 ? void 0 : _a.throwIfAborted();
            yield abortOr(options === null || options === void 0 ? void 0 : options.signal, this.createLevelDatabase());
            return abortOr(options === null || options === void 0 ? void 0 : options.signal, this.db.put(String(key), value));
        });
    }
    delete(key, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            (_a = options === null || options === void 0 ? void 0 : options.signal) === null || _a === void 0 ? void 0 : _a.throwIfAborted();
            yield abortOr(options === null || options === void 0 ? void 0 : options.signal, this.createLevelDatabase());
            return abortOr(options === null || options === void 0 ? void 0 : options.signal, this.db.del(String(key)));
        });
    }
    isEmpty(options) {
        var e_3, _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                for (var _b = __asyncValues(this.keys(options)), _c; _c = yield _b.next(), !_c.done;) {
                    const _key = _c.value;
                    return false;
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
            return true;
        });
    }
    clear() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.createLevelDatabase();
            yield this.db.clear();
            yield this.compactUnderlyingStorage();
        });
    }
    batch(operations, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            (_a = options === null || options === void 0 ? void 0 : options.signal) === null || _a === void 0 ? void 0 : _a.throwIfAborted();
            yield abortOr(options === null || options === void 0 ? void 0 : options.signal, this.createLevelDatabase());
            return abortOr(options === null || options === void 0 ? void 0 : options.signal, this.db.batch(operations));
        });
    }
    compactUnderlyingStorage(options) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            (_a = options === null || options === void 0 ? void 0 : options.signal) === null || _a === void 0 ? void 0 : _a.throwIfAborted();
            yield abortOr(options === null || options === void 0 ? void 0 : options.signal, this.createLevelDatabase());
            const range = this.sublevelRange;
            if (!range) {
                return;
            }
            // additional methods are only available on the root API instance
            const root = this.root;
            if (root.db.supports.additionalMethods.compactRange) {
                return abortOr(options === null || options === void 0 ? void 0 : options.signal, (_c = (_b = root.db).compactRange) === null || _c === void 0 ? void 0 : _c.call(_b, ...range));
            }
        });
    }
    get sublevelRange() {
        const prefix = this.db.prefix;
        if (!prefix) {
            return undefined;
        }
        // use the separator to derive an exclusive `end` that will never match to a key (which matches how `abstract-level` creates a `boundary`)
        return [prefix, prefix.slice(0, -1) + String.fromCharCode(prefix.charCodeAt(prefix.length - 1) + 1)];
    }
    get root() {
        let db = this.db;
        for (const parent = db.db; parent && parent !== db;) {
            db = parent;
        }
        return new LevelWrapper(this.config, db);
    }
    createLevelDatabase() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            (_a = this.db) !== null && _a !== void 0 ? _a : (this.db = yield this.config.createLevelDatabase(this.config.location, {
                keyEncoding: 'utf8',
                valueEncoding: this.config.valueEncoding
            }));
        });
    }
    dump() {
        var e_4, _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.db) {
                return;
            }
            try {
                for (var _b = __asyncValues(this.db.iterator()), _c; _c = yield _b.next(), !_c.done;) {
                    const [key, value] = _c.value;
                    console.debug(key, value);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield _a.call(_b);
                }
                finally { if (e_4) throw e_4.error; }
            }
        });
    }
}
