// @tbd54566975/dwn-sdk-js/dist/cjs/src/interfaces/messages/handlers/messages-get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessagesGetHandler = void 0;
var data_stream_js_1 = require("../../../utils/data-stream.js");
var dwn_constant_js_1 = require("../../../core/dwn-constant.js");
var encoder_js_1 = require("../../../utils/encoder.js");
var message_reply_js_1 = require("../../../core/message-reply.js");
var messages_get_js_1 = require("../messages/messages-get.js");
var auth_js_1 = require("../../../core/auth.js");
var message_js_1 = require("../../../core/message.js");
var MessagesGetHandler = /** @class */ (function () {
    function MessagesGetHandler(didResolver, messageStore, dataStore) {
        this.didResolver = didResolver;
        this.messageStore = messageStore;
        this.dataStore = dataStore;
    }
    MessagesGetHandler.prototype.handle = function (_a) {
        var tenant = _a.tenant, message = _a.message;
        return __awaiter(this, void 0, void 0, function () {
            var messagesGet, e_1, e_2, promises, messageCids, _loop_1, this_1, messageCids_1, messageCids_1_1, messageCid, messages, messages_1, messages_1_1, entry, message_1, _b, messageInterface, method, dataCid, dataSize, messageCid, result, dataBytes, e_3_1;
            var e_4, _c, e_3, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _e.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, messages_get_js_1.MessagesGet.parse(message)];
                    case 1:
                        messagesGet = _e.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _e.sent();
                        return [2 /*return*/, message_reply_js_1.MessageReply.fromError(e_1, 400)];
                    case 3:
                        _e.trys.push([3, 6, , 7]);
                        return [4 /*yield*/, (0, auth_js_1.authenticate)(message.authorization, this.didResolver)];
                    case 4:
                        _e.sent();
                        return [4 /*yield*/, (0, auth_js_1.authorize)(tenant, messagesGet)];
                    case 5:
                        _e.sent();
                        return [3 /*break*/, 7];
                    case 6:
                        e_2 = _e.sent();
                        return [2 /*return*/, message_reply_js_1.MessageReply.fromError(e_2, 401)];
                    case 7:
                        promises = [];
                        messageCids = new Set(message.descriptor.messageCids);
                        _loop_1 = function (messageCid) {
                            var promise = this_1.messageStore.get(tenant, messageCid)
                                .then(function (message) {
                                return { messageCid: messageCid, message: message };
                            })
                                .catch(function (_) {
                                return { messageCid: messageCid, message: undefined, error: "Failed to get message ".concat(messageCid) };
                            });
                            promises.push(promise);
                        };
                        this_1 = this;
                        try {
                            for (messageCids_1 = __values(messageCids), messageCids_1_1 = messageCids_1.next(); !messageCids_1_1.done; messageCids_1_1 = messageCids_1.next()) {
                                messageCid = messageCids_1_1.value;
                                _loop_1(messageCid);
                            }
                        }
                        catch (e_4_1) { e_4 = { error: e_4_1 }; }
                        finally {
                            try {
                                if (messageCids_1_1 && !messageCids_1_1.done && (_c = messageCids_1.return)) _c.call(messageCids_1);
                            }
                            finally { if (e_4) throw e_4.error; }
                        }
                        return [4 /*yield*/, Promise.all(promises)];
                    case 8:
                        messages = _e.sent();
                        _e.label = 9;
                    case 9:
                        _e.trys.push([9, 16, 17, 18]);
                        messages_1 = __values(messages), messages_1_1 = messages_1.next();
                        _e.label = 10;
                    case 10:
                        if (!!messages_1_1.done) return [3 /*break*/, 15];
                        entry = messages_1_1.value;
                        message_1 = entry.message;
                        if (!message_1) {
                            return [3 /*break*/, 14];
                        }
                        _b = message_1.descriptor, messageInterface = _b.interface, method = _b.method;
                        if (messageInterface !== message_js_1.DwnInterfaceName.Records || method !== message_js_1.DwnMethodName.Write) {
                            return [3 /*break*/, 14];
                        }
                        dataCid = message_1.descriptor.dataCid;
                        dataSize = message_1.descriptor.dataSize;
                        if (!(dataCid !== undefined && dataSize <= dwn_constant_js_1.DwnConstant.maxDataSizeAllowedToBeEncoded)) return [3 /*break*/, 14];
                        return [4 /*yield*/, message_js_1.Message.getCid(message_1)];
                    case 11:
                        messageCid = _e.sent();
                        return [4 /*yield*/, this.dataStore.get(tenant, messageCid, dataCid)];
                    case 12:
                        result = _e.sent();
                        if (!result) return [3 /*break*/, 14];
                        return [4 /*yield*/, data_stream_js_1.DataStream.toBytes(result.dataStream)];
                    case 13:
                        dataBytes = _e.sent();
                        entry.encodedData = encoder_js_1.Encoder.bytesToBase64Url(dataBytes);
                        _e.label = 14;
                    case 14:
                        messages_1_1 = messages_1.next();
                        return [3 /*break*/, 10];
                    case 15: return [3 /*break*/, 18];
                    case 16:
                        e_3_1 = _e.sent();
                        e_3 = { error: e_3_1 };
                        return [3 /*break*/, 18];
                    case 17:
                        try {
                            if (messages_1_1 && !messages_1_1.done && (_d = messages_1.return)) _d.call(messages_1);
                        }
                        finally { if (e_3) throw e_3.error; }
                        return [7 /*endfinally*/];
                    case 18: return [2 /*return*/, {
                            status: { code: 200, detail: 'OK' },
                            messages: messages
                        }];
                }
            });
        });
    };
    return MessagesGetHandler;
}());
exports.MessagesGetHandler = MessagesGetHandler;
